### DISCUSSION TITLE
How to import complex geometry like an airfoil in .i3d file ?

### DISCUSSION THREAD

**Discussion by gauravity30** [xcompact3d/Incompact3d#192] [Q&A] [answered]
I have some airfoils over which i want to run the simulations. How do I import the airfoil geometry for simulation ? One of the way I found was to replace the equation of cylinder with the equation of the geometry in the BC_Cylinder.f90 file. Is there any other way to import the geometry like using the coordinates of the airfoil.

**Comment by slaizet** [comment, marked as answer]
Hi,
We are going to prepare a video tutorial on how to do import a stl file in Xcompact3d (it is possible via the Sandbox). In the meantime, using equations for your airfoil is the best option :-)
Best,
Sylvain

**Comment by gauravity30** [comment]
Thank you ! It will be very helpful.

**Comment by gauravity30** [comment]
Hello @slaizet !
I tried using the sandbox mode through the python code xcompact3d-toolbox. I was facing trouble installing the library. The dependencies of the library were not supporting each other. So, I could not use the option to import STL using that. 

So, with some reading i developed my own python code to produce Epsilon function for any STL geometry. Currently,  I am trying to input the geometry in cylinder case, by reading the indices of the points inside geometry and setting `epsi(x,y,z) = remp` for those coordinates. 
(Check out the python code I used, [here](https://github.com/airwarriorg91/xcompact3d-stl_2_epsilon_generator))
```
if (nxi.eq.1.and.nxf.eq.nx.and.&
         nyi.eq.xstart(2).and.nyf.eq.xend(2).and.&
        nzi.eq.xstart(3).and.nzf.eq.xend(3)) then
         open(2, file = 'src/epsilon.dat', status = 'unknown')
         read(2,*) len
         allocate(ex(len),ey(len),ez(len))
         read(2,*) (ex(i), ey(i), ez(i), i=1, len)
         close(2)
         do i=1,len
            epsi(ex(i),ey(i),ez(i)) = remp
         enddo
      endif
```
This is the modification I made in BC-Cylinder.f90. The simulation starts but all the values are NaN. Can you please me how to input the geometry correctly from a epsilon file ? Do we need to write an extra subroutines for this ?

**Reply by slaizet** [reply]
Dear Gaurav,

You should not modify BC-Cylinder.f90. The file of interest for you is genepsi3d.f90. This is where we prepare all the files and arrays for the immersed boundary method approach, in particular the subroutine geomcomplex_io.

Best,

Sylvain

**Comment by gauravity30** [comment]
Dear Sylvain,
Thank you for the reply I actually looked in the genepsi3d.f90 and subroutine geomcomplex_io. I was a little confused. If I opt for reading the epsilon file using the geomcomplex_io subroutine, then I have to read the files for the other parameters `nobjx,nobjy,nobjz,xi,xf,yi,yf,zi,zf, nxipif,nxfpif,nyipif,nyfpif,nzipif,nzfpif,nobjmax,npif`. Can you tell what these parameters are ? I couldn't understand what these arrays actually mean.

Interestingly, the code works with my modification if I set the IBM method as forced boundary method (1) instead of method mentioned in "A DNS study of jet control with microjets using an alternating direct on forcing strategy". I am testing whether I am getting good results with the forced boundary method. Also, I need to check if it works with the other method. 

Regards,
Gaurav

**Reply by slaizet** [reply]
Hi,
iibm=1 correspond to an IBM with no reconstruction inside the immersed object. iibm=2 correspond to the IBM described in Gautier et al., 2014, with flow reconstructions in each spatial direction for each of the velocity component, hence why you need all those arrays for the reconstruction.
Best,
Sylvain

**Reply by gauravity30** [reply]
Hello,
Okay, understood. Thank you for the explanation. Are the arrays described in the paper ? 
Thanks,
Gaurav

**Reply by gauravity30** [reply]
Hello Sylvain,
From reading the genepsi3d.f90 and the genepsi.py of xcompact-toolbox library, I understood what `nobjx, nobjy, nobjz` and `xi, xf, yi, yf, zi, zf` mean. The `nobj` represents the number of objects (like a cylinder or cube) in each of the direction and `xi, xf, ..` represent the boundaries of the objects in each of the direction. But I couldn't understand what `nxifpif, nyifpif and nzipif` mean. Can you please explain what they are ? Please correct me if I am wrong anywhere.

Thank you !
Gaurav

**Reply by slaizet** [reply]
Hi, 
`nxifpif, nyifpif and nzipif` are arrays needed for the reconstruction of the velocity inside the immersed objets with Lagrange polynomials. It more or less correspond to the number of mesh nodes in the fluid needed to reconstruct the velocity. As an indication, we typically use two fluid mesh nodes for the Lagrange polynomials, and we skip the fluid mesh node closest to the immersed object.
Best,
Sylvain

**Reply by gauravity30** [reply]
Hii,
Thanks a lot for explaining that. 

Thanks,
Gaurav

**Reply by gauravity30** [reply]
Hello @slaizet,
I was going through the `genepsi3d.f90` file. I am little confused in the part where `nobjy` and `nobjz` are computed. What is the call `transpose_x_to_y(ep1,ep2)` and `transpose_y_to_z(ep2,ep3)`. I tried checking what it was doing so I tried printing the size of the arrays `ep1, ep2, ep3`. The shape of the array is constant and not changing, so what the calls are doing ? I also noticed these calls in `verif_epsi` subroutine while generating `nyipif, nyfpif, ..`. I apologize for asking so many questions. I am trying to write the code in python, so I am trying to understand the inner working. Thanks a lot ! 

Thanks,
Gaurav

**Reply by slaizet** [reply]
Hi, 
As you should know, all 3D arrays ending with 1, 2 or 3 are 3D arrays in X-, Y- or Z- pencils. They do not have the same sizes. The transpose subroutines are tools to switch an array from one pencil decomposition to another. Please have a read at the relevant publications for more information.
Best,
Sylvain

**Reply by gauravity30** [reply]
Hii,
Thanks for the explanation. So, the transpose subroutines are relevant to the domain decomposition process. Since, I will be generating the `nobjx, nobjy, ..` etc. in python without any domain decomposition. I think I wouldn't require to implement the transpose subroutines. Does, the `geomcomplex_io` subroutine handle this domain decomposition after it reads the data from the .dat files ?

Thanks,
Gaurav

**Reply by slaizet** [reply]
Hi,
I am not quite sure what you are trying to do here (doing in Python what is currently being done in Fortran?). If you are trying to generate in Python all the arrays needed for the reconstruction for the IBM then yes you do not need to worry about the transpose subroutines. 
Best,
Sylvain

**Reply by gauravity30** [reply]
Hi,
Thanks a lot for the clarification. Yes, I am trying to generate the arrays in Python for reconstruction of IBM. 

Thanks,
Gaurav
